README.md
# Installation

To install, run the following lines: 
```
install.packages("devtools")
devtools::install_github("openanalytics/poissontris")
```

# Running the game

```
library(poissontris)
runPoissontris()
```

Once the app is running, you can click on the 'Show PC Controls' button for keyboard commands. 
Optionally, use a program (e.g. reWASD in Windows) to map a controller to those commands. 


inst/NEWS
0.0.3
    o implemented wallkicks
    o fixed a minor plotting bug on game reset
0.0.2
    o stacked svg elements for improved performance
    o eliminated all button observers
0.0.1
    o initial alpha version


inst/ui/server.R


function(input, output, session) {
  
  # game states
  gameOver <- FALSE
  gameActive <- FALSE
  
  # how many frames to hold a button before it fires again
  holdDuration <- 4
  
  # can the down key be pressed (sets to false on line clear, true on keyup)
  canPressDown <- TRUE
  
  # duration of button press (how many frames has it been held)
  leftDuration <- 0
  rightDuration <- 0
  downDuration <- 0
  clockDuration <- 0
  counterDuration <- 0
  noDownCounter <- 0
  
  # old counter values
  leftCounterOld <- 0
  downCounterOld <- 0
  rightCounterOld <- 0
  clockCounterOld <- 0
  counterCounterOld <- 0
  
  # game data
  lambda <- 3.5
  score <- 0
  pieceMoved <<- TRUE
  
  # plot reactive value
  rv <- reactiveValues()
  
  # debugging
#  tickTocks <- 0
#  oldTime <- Sys.time()
  
  ##########################
  # define reactive values #
  nextPiece <- reactiveValues(shape = makeShape(lambda))
  observeEvent(values$numShapes, {
        nextPiece$shape <- makeShape(lambda)
      })
  
  values <- reactiveValues(yOffset = 20, 
      shape = makeShape(lambda),
      xOffset = 5, 
      bg = makeBg(), 
      numShapes = 0,
      highScore = if (file.exists("highScore.rds")) readRDS("highScore.rds") else 0)
  
  
  ######################
  #    baseline plot   #
  s1 <- svgstring(standalone=FALSE, height=600/72, width=297/72)
  par(mar = c(0, 0, 0, 0))
  blankPlot(xlim = c(-0.5, 10.5), 
      ylim = c(-0.5, 22.5))   # , asp = 1
  rect(xleft = 0, xright = 10,
      ybottom = 20, ytop = 22, 
      lwd = 4, border = NA, col = grey(0.95))
  for(i in 1:9)        # TODO replace 
    segments(x0 = i, y0 = 0, y1 = 22, 
        col = gray(0.7))
  segments(x0 = 0, x1 = 10, y0 = 0, lwd = 4)
  segments(x0 = 0, y0 = 0, y1 = 22, lwd = 4)
  segments(x0 = 10, y0 = 0, y1 = 22, lwd = 4)
  dev.off()
  output[["s1"]] <- renderUI({ HTML(s1()) })
  
  
  
  ############################
  # plots and value updating #
  ms <- 33
  tickTock30 <- reactiveTimer(ms)
  observeEvent(tickTock30(), {
        
#        tickTocks <<- tickTocks + 1
#        if(tickTocks %% 30 == 0) {
#          newTime <- Sys.time()
#          print(as.numeric(difftime(newTime, oldTime)))
#          oldTime <<- newTime
#        }

        # TODO think about using the %% mod functions again
        
        if(gameActive) {
          
          
          # left movement
          if(input$leftCounter == leftCounterOld & input$left == 50) {
            leftDuration <<- leftDuration + 1
          } else 
            leftDuration <<- 0
          
          if(input$leftCounter > leftCounterOld | 
              (leftDuration > holdDuration)) {  
            if(!detectCollision(values, xChange = -1)) {
              values$xOffset <- values$xOffset - 1
              pieceMoved <<- TRUE
            }
            leftCounterOld <<- input$leftCounter
          }
          
          # right movement
          if(input$rightCounter == rightCounterOld & input$right == 50) {
            rightDuration <<- rightDuration + 1
          } else 
            rightDuration <<- 0
          
          if(input$rightCounter > rightCounterOld | 
              (rightDuration > holdDuration)) {  
            if(!detectCollision(values, xChange = +1)){
              values$xOffset <- values$xOffset + 1
              pieceMoved <<- TRUE
            }
            rightCounterOld <<- input$rightCounter
            
          }
          
          
          # clockwise rotation
          if(input$clock > clockCounterOld) {
            out <- detectCollision(values, rotateClockwise = TRUE)
            if(!out[[1]]) {
              if(length(out) > 1)
                values$xOffset <- values$xOffset + out$xChange
              values$shape <- rotateClockwise(values$shape)
              
              clockCounterOld <<- input$clock
              pieceMoved <<- TRUE
            }
          }
          
          
          # counter rotation
          if(input$counter > counterCounterOld){
            out <- detectCollision(values, rotateCounter = TRUE)
            if(!out[[1]]) {
              if(length(out) > 1)
                values$xOffset <- values$xOffset + out$xChange
              values$shape <- rotateCounter(values$shape) 
              
              counterCounterOld <<- input$counter
              pieceMoved <<- TRUE
            }
          }
          
          # the down key is not pressed
          if(input$down == 30)
            canPressDown <<- TRUE
          
          # down movement
          noDownCounter <<- noDownCounter + 1
          if(input$downCounter == downCounterOld & input$down == 50) {
            downDuration <<- downDuration + 1
          } else 
            downDuration <<- 0
          
          if(noDownCounter >= 450/ms | (canPressDown & (input$downCounter > downCounterOld)) | 
              (canPressDown & (downDuration > holdDuration))) {   #& downDuration %% holdDuration == 1
            pieceMoved <<- TRUE
            
            downCounterOld <<- input$downCounter
            if(detectCollision(values, yChange = -1)) { 
              
              canPressDown <<- FALSE
              
              # update values          
              values <- addPieceToBg(values, lambda, nextPiece)
              swapAvailable <<- 1
              
              # clear lines and change score
              linesCleared <- clearLines(values)
              if(linesCleared > 0) {
                score <<- score + lambda + (linesCleared-1)*2*lambda
                lambda <<- lambda + 0.25 * linesCleared
              }
              values$numShapes <- values$numShapes + 1
              
              # plot the blocks that have already fallen
              s2 <- svgstring(standalone=FALSE, height=600/72, width=297/72, 
                  bg = "transparent")
              par(mar = c(0, 0, 0, 0))
              blankPlot(xlim = c(-0.5, 10.5), 
                  ylim = c(-0.5, 22.5))   # , asp = 1
              plotBg(values$bg)
              dev.off()
              rv$s2 <- s2() 
              
              # check for game end
              if(any(values$bg$y > 19)) {
                gameOver <<- TRUE
                gameActive <<- FALSE
              }
              
              
              
              
            } else {
              # move the piece if no collision
              values$yOffset <- values$yOffset - 1
            }
            downWasPressed <<- FALSE
            noDownCounter <<- 0
            
          }
        }
        
        if(pieceMoved) {
          # plot the active block
          s3 <- svgstring(standalone=FALSE, height=600/72, width=297/72, 
              bg = "transparent")
          par(mar = c(0, 0, 0, 0))
          blankPlot(xlim = c(-0.5, 10.5), 
              ylim = c(-0.5, 22.5))   # , asp = 1
          plotShape(values$shape, xOffset = values$xOffset,
              yOffset = values$yOffset)
          dev.off()
          rv$s3 <- s3() 
          pieceMoved <<- FALSE
        }
        
        if(!gameActive) {
          s4 <- svgstring(standalone=FALSE, height=600/72, width=297/72, 
              bg = "transparent")
          par(mar = c(0, 0, 0, 0))
          blankPlot(xlim = c(-0.5, 10.5), 
              ylim = c(-0.5, 22.5))   # , asp = 1
          rect(xleft = 1.5, xright = 8.5,
              ybottom = 10, ytop = 14, 
              lwd = 4, col = "white")
          text(x = 5, y = 13, "Press Start to Begin")
          if(gameOver) {
            text(x = 5, y = 11, paste0("Score: ", score))
            if (score >= values$highScore) {
              saveRDS(score, "highScore.rds")
              values$highScore <- score
              text(x = 5, y = 12, "High Score!", col = "red")
            }
          }
          
          dev.off()
          rv$s4 <- s4() 
        } 
        
      })
  
  
  # ----------- other layered plots ------------- #
  
  output[["s2"]] <- renderUI({ HTML(rv$s2) })
  output[["s3"]] <- renderUI({ HTML(rv$s3) })
  output[["s4"]] <- renderUI({ HTML(rv$s4) })
  
  
  
  observeEvent(nextPiece$shape, {
        s <- svgstring(standalone=FALSE, height=170/72, width=170/72)
        par(mar = c(0, 0, 0, 0))
        valVec <- c(nextPiece$shape$x, nextPiece$shape$y)
        valRange <- c(min(-3.5, -1*max(abs(valVec))-0.5),
            max(max(abs(valVec))+1.5, 4.5) )
        
        blankPlot(xlim = valRange, ylim = valRange)   # , asp = 1
        rect(valRange[1], valRange[1],
            valRange[2], valRange[2], lwd = 4, col = NA, bg = NA)
        plotShape(nextPiece$shape, xOffset = 0,
            yOffset = 0)
        dev.off()
        rv$nextPiece <- s() 
      })
  output[["nextPiece"]] <- renderUI({ HTML(rv$nextPiece) })
  
  observeEvent(values$shape, {
        s <- svgstring(standalone=FALSE, height=300/72, width=450/72)
        par(mar = c(0, 0, 0, 0))
        blankPlot(xlim = c(0, 11), ylim = c(-0.07, 0.25))
        rect(0, -0.07, 11, 0.25, lwd = 4)
        numBlocks <- length(values$shape$y)
        
        for(i in 2:9) {
          rectCol <- ifelse(i == numBlocks, values$shape$col, "white")
          rect(xleft = i-0.5, ybottom = 0, xright=i+0.5, ytop = dpois(i, lambda), 
              lwd = 3, col = rectCol)
        }
        
        rectCol <- ifelse(numBlocks %in% c(0, 1), values$shape$col, "white")
        rect(xleft = 0.5, ybottom = 0, xright=1.5, 
            ytop = ppois(1, lambda), 
            lwd = 3, col = rectCol)
        
        rectCol <- ifelse(numBlocks > 9, values$shape$col, "white")
        rect(xleft = 10-0.5, ybottom = 0, xright=10+0.5, 
            ytop = ppois(10, lambda, lower.tail = FALSE), 
            lwd = 3, col = rectCol)
        
        text(x = c(1:9, 10.1), y = rep(-0.01, 10), labels = c("0-1", 2:9, "10+"), 
            font = 2, adj = c(0.5, 1), cex = 1.2)
        text(x = 5.5, y = -0.04, labels = "Piece Size (# Blocks)", 
            font = 1, adj = c(0.5, 1), cex = 1.4)
        dev.off()
        rv$pdf <- s() 
      })
  output[["pdf"]] <- renderUI({ HTML(rv$pdf) })
  
  
  sControl <- svgstring(standalone=FALSE, height=170/72, width=170/72)
  par(mar = c(0, 0, 0, 0))
  blankPlot(xlim = c(-0.6, 1), ylim = c(-0.3, 1.3))
  points(x = c(0.5, 0.75, 0.5, 0.25), y = c(0.25, 0.5, 0.75, 0.5), 
      pch = 21, bg = oaColors(c("green", "red", "yellow", "cyan")), 
      cex = 4)
  
  text(x = 0.5, y = 0.95, adj = c(0.5, 0), "Swap Pieces", font = 2, 
      cex = 0.8)
  text(x = 0, y = 0.5, adj = c(1, 0.5), "Rotate \nCounter-\nClockwise",
      font = 2, cex = 0.8)
  text(x = 0.5, y = 0.06, adj = c(0.5, 1), 
      "Rotate \nClockwise", font = 2, cex = 0.8)
  rect(-0.6, -0.3, 1, 1.3, lwd = 4)
  dev.off()
  output[["controls"]] <- renderUI({ HTML(sControl()) })
  
  
  # ------------  game functions --------------- #
  
  observeEvent(input$startGame, {
        gameActive <<- TRUE
        
        # remove the start game box
        s4 <- svgstring(standalone=FALSE, height=600/72, width=297/72, 
            bg = "transparent")
        par(mar = c(0, 0, 0, 0))
        blankPlot(xlim = c(-0.5, 10.5), 
            ylim = c(-0.5, 22.5))   # , asp = 1
        dev.off()
        rv$s4 <- s4() 
        
        if(gameOver) {     
          
          lambda <<- 3.5
          score <<- 0
          
          values$yOffset <- 20
          values$shape <- makeShape(lambda)
          values$xOffset <- 5
          values$bg = makeBg() 
          values$numShapes = 0
          
          nextPiece$shape <- makeShape(lambda)
          
          # reset the fallen blocks
          s2 <- svgstring(standalone=FALSE, height=600/72, width=297/72, 
              bg = "transparent")
          par(mar = c(0, 0, 0, 0))
          blankPlot(xlim = c(-0.5, 10.5), 
              ylim = c(-0.5, 22.5))   # , asp = 1
          plotBg(values$bg)
          dev.off()
          rv$s2 <- s2()
          
          # draw the new active piece
          s3 <- svgstring(standalone=FALSE, height=600/72, width=297/72, 
              bg = "transparent")
          par(mar = c(0, 0, 0, 0))
          blankPlot(xlim = c(-0.5, 10.5), 
              ylim = c(-0.5, 22.5))   # , asp = 1
          plotShape(values$shape, xOffset = values$xOffset,
              yOffset = values$yOffset)
          dev.off()
          rv$s3 <- s3() 
          
          values$highScore <- if (file.exists("highScore.rds")) readRDS("highScore.rds") else 0
          gameOver <<- FALSE
          
        }
        
      })
  
  swapAvailable <- 1
  observeEvent(input$swap, {
        if(swapAvailable) {
          swapAvailable <<- 0
          tmpShp <- nextPiece$shape
          nextPiece$shape <- values$shape
          values$shape <- tmpShp
          values$xOffset <- 4
          values$yOffset <- 20
        }
        
      })
  
  
  
  # ----------- score output ----------------- #
  
  observeEvent(values$numShapes, {
        output$lambda <- renderText(lambda)
        output$score <- renderText(score)
      })
  
  output$highScore <- renderText(values$highScore)
  
  output$pdfTitle <- renderText(paste0("Poisson PDF (&lambda; = ",
          values$shape$lambda[1], ")"))
  
  
}

R/gameFunctions.R
#' Detects whether a proposed change in the active piece will cause a collision in 
#' the background
#' @param values reactive values object with current piece and bg information
#' @param xChange proposed x location change
#' @param yChange proposed y location change
#' @param rotateClockwise (logical) whether to rotate the piece clockwise
#' @param rotateCounter  (logical) whether to rotate the piece counter clockwise
#' @param tryWallkick (logical) specify whether the function should attempt xShifts
#'    to resolve a rotation
#' @return (logical) whether a collision will occur
#' @author Jason Waddell
#' @export 
detectCollision <- function(values, xChange = 0, yChange = 0, 
    rotateClockwise = FALSE, rotateCounter = FALSE, tryWallkick = TRUE) {
  
  out <- FALSE
  
  tmpShape <- values$shape
  if(rotateClockwise)
    tmpShape <- rotateClockwise(tmpShape)
  if(rotateCounter)
    tmpShape <- rotateCounter(tmpShape)
  
  tmpShape$x <- tmpShape$x + values$xOffset + xChange
  tmpShape$y <- tmpShape$y + values$yOffset + yChange
  
  # shape beneath the floor
  if(min(tmpShape$y) < 0) 
    return(TRUE)
  
  # shape beyond the right wall
  if(max(tmpShape$x) > 9) {
    if(tryWallkick & (rotateClockwise | rotateCounter)) 
      for(i in (-1):(-3)) {
        collision <- detectCollision(values, xChange = i,
            rotateClockwise = rotateClockwise, tryWallkick = FALSE)
        if(!collision)
          return(list(FALSE, xChange = i))
      }
    
    return(TRUE)   
  }
  
  # shape beyond the left wall
  if(min(tmpShape$x) < 0) {
    if(tryWallkick & (rotateClockwise | rotateCounter))   
      for(i in 1:3) {
        collision <- detectCollision(values, xChange = i,
            rotateClockwise = rotateClockwise, tryWallkick = FALSE)
        if(!collision)
          return(list(FALSE, xChange = i))
      }
    
    return(TRUE)
  }
  
  
  # check collision with background pieces
  if(length(values$bg$y) > 0) 
    if((min(tmpShape$y) <= max(values$bg$y))) {
      
      idx <- which(values$bg$y >= min(tmpShape$y) & 
              values$bg$x %in% unique(tmpShape$x)) 
      
      z1 <- paste(tmpShape$x, tmpShape$y, sep = ",")
      z2 <- paste(values$bg$x[idx], values$bg$y[idx], sep = ",")
      
      # collision detected with background, but let's try wallkicks
      if(any(z1 %in% z2)) {
        if(tryWallkick & (rotateClockwise | rotateCounter))   
          for(i in c(1, -1, 1, 2)) {
            collision <- detectCollision(values, xChange = i,
                rotateClockwise = rotateClockwise, tryWallkick = FALSE)
            if(!collision)
              return(list(FALSE, xChange = i))
          }
        
        # no wallkicks worked
        return(TRUE)
      }
        
    }
  
  return(FALSE)
  
}


#' Check if any lines need to be cleared, and update objects accordingly
#' @param values reactive values object with current piece and bg information
#' @return (numeric) number of lines cleared
#' @author Jason Waddell
#' @export 
clearLines <- function(values) {
  
  yTable <- table(values$bg$y)  
  if(max(yTable) == 10) {
    remLines <- as.numeric(names(which(yTable == 10)))
    if(length(values$bg$y) == (length(remLines) * 10)) {
      values$bg <- makeBg()
    } else {
      idx <- which(values$bg$y %in% remLines)
      values$bg$y <- values$bg$y[-idx]
      values$bg$x <- values$bg$x[-idx]
      values$bg$col <- values$bg$col[-idx]
      
      for(i in 1:length(values$bg$y)) {
        values$bg$y[i] <- 
            values$bg$y[i] - sum(remLines < values$bg$y[i])
      }
      
    }
    
    return(length(remLines))
  }
  return(0)  # return TRUE for altering fal intervall
}



#' Create a blank background list
#' @return the background list
#' @author Jason Waddell
#' @export 
makeBg <- function() {  
  bg <- list(x=numeric(), y=numeric(), col=character())
  return(bg)  
}


#' When generating a new piece, find candidate block locations
#' @param pieceMat current piece data.frame
#' @return list of candidate locations
#' @author Jason Waddell
#' @export 
findOpenLocations <- function(pieceMat){
  ll <- list()
  for(i in 1:nrow(pieceMat))
    ll[[i]] <- adjacentLocations(pieceMat[i,])
  
  ll <- do.call("rbind.data.frame", ll)
  ll <- ll[!duplicated(ll),] # to find the unique rows
  
  ll <- ll[!duplicated(rbind.data.frame(ll, pieceMat), 
          fromLast = TRUE)[1:nrow(ll)],]
  
  return(ll)
}

#' Find adjacent locations to a given block
#' @param pieceRow One row of the current piece data.frame
#' @return matrix of adjacent locations
#' @author Jason Waddell
#' @export 
adjacentLocations <- function(pieceRow){
  x <- as.numeric(pieceRow[,1]); y <- as.numeric(pieceRow[,2])
  xOut <- c(x, x+1, x, x-1)
  yOut <- c(y+1, y, y-1, y)
  out <- cbind(xOut, yOut); colnames(out) <- c("x", "y")
  return(out)
}

#' Make a randomly generated piece, sampled from the current lambda value
#' @param lambda value of lambda (to be handed to rpois)
#' @return a piece list object
#' @author Jason Waddell
#' @export 
makeShape <- function(lambda=3.5){
  
  colVec <- oaPalette(9)
  myCol <- sample(colVec, 1)
  
  value <- rpois(1, lambda)
  n <- max(value, 1)
  
  x <- 1; y <- 1; pieceMat <- cbind.data.frame(x, y)
  
  if(n > 1)
    for(i in 2:n){
      openSpaces <- findOpenLocations(pieceMat)
      nextSpace <- openSpaces[sample(1:nrow(openSpaces), size = 1),]
      pieceMat <- rbind.data.frame(pieceMat, nextSpace)
    }		
  
  pieceMat$x <- pieceMat$x - round(median(pieceMat$x))
  pieceMat$y <- pieceMat$y - round(median(pieceMat$y))
  pieceMat <- cbind.data.frame(pieceMat, col = rep(myCol, n), 
      value = rep(value, n), 
      lambda = rep(lambda, n))
  pieceMat$col <- as.character(pieceMat$col)
  pieceMat <- as.list(pieceMat)
  
  return(pieceMat)
}




#' Add the current piece to the background (because of a collision)
#' @param values reactive values object with current piece and bg information 
#' @param lambda  current lambda value
#' @param nextPiece the reactive value object for the next piece (to become the active piece)
#' @return values reactive values object with current piece and bg information
#' @author Jason Waddell
#' @export 
addPieceToBg <- function(values, lambda, nextPiece) {
  values$shape$x <- values$shape$x + values$xOffset
  values$shape$y <- values$shape$y + values$yOffset
  
  values$bg$x <- c(values$bg$x, values$shape$x)
  values$bg$y <- c(values$bg$y, values$shape$y)
  values$bg$col <- c(values$bg$col, values$shape$col)
  values$shape <- nextPiece$shape
  values$xOffset <- 4
  values$yOffset <- 20
  return(values)
}



#' Rotate a piece clockwise
#' @param shape the shape list object
#' @return updated shape list object
#' @author Jason Waddell
#' @export 
rotateClockwise <- function(shape) {
  
  tmpX <- shape$x
  shape$x <- shape$y
  shape$y <- -1 * tmpX
  return(shape)
}

#' Rotate a piece counterclockwise
#' @param shape the shape list object
#' @return updated shape list object
#' @author Jason Waddell
#' @export 
rotateCounter <- function(shape) {
  
  tmpY <- shape$y
  shape$y <- shape$x
  shape$x <- -1 * tmpY
  return(shape)
}


R/runPoissontris.R
#' Run the Cardiovascular MFR shiny application
#' @return no return value
#' @importFrom shiny runApp
#' @export
runPoissontris <- function(){
	
	# (1) Copy the UI files & folders from "inst/ui" for local use
#	tmpDir <- tempdir()
#	setwd(tmpDir)
	library(shiny)
	uiDir <- system.file("ui", package = "poissontris")
  setwd(uiDir)
#	uiFiles <- list.files(path = uiDir, full.names = FALSE, recursive = TRUE)
#	
#	sapply(uiFiles, function(from) {
#				
#				to <- file.path(tmpDir, from)
#				toDir <- dirname(to)
#				
#				if (!dir.exists(toDir)) {
#					
#					dir.create(path = toDir, recursive = TRUE)
#					
#				}
#				
#				file.copy(from = file.path(uiDir, from), to = to, overwrite = TRUE)
#				
#			})  
	
	
	# (2) Run the application
	runApp()
	
}






R/plottingFunctions.R



#' Plot the current shape list object
#' @param shape shape list object containing x, y and color values
#' @param xOffset current x offset modifying shape object
#' @param yOffset current y offset modifying shape object
#' @return none (objects plotted to current device)
#' @author Jason Waddell
#' @export 
plotShape <- function(shape, xOffset, yOffset){
  if(length(shape$y) > 0)  
    rect(xleft = shape$x+xOffset,
          xright = shape$x+1+xOffset, 
          ybottom = shape$y+yOffset,
          ytop = shape$y+1+yOffset, 
          col = shape$col, 
          border = "black", lwd = 2)
}


#' Plot the background list object
#' @param bg list containing x, y and color values
#' @return none (objects plotted to current device)
#' @author Jason Waddell
#' @export 
plotBg <- function(bg){
  
  if(length(bg) > 0)
    rect(xleft = bg$x,
              xright = bg$x+1, 
              ybottom = bg$y,
              ytop = bg$y+1, 
              col = bg$col, 
              border = "black", lwd = 2)
  
}

##' Make a plot displaying number of blocks in piece versus lambda
##' @param record the record reactive value object
##' @return none (objects plotted to current device)
##' @author Jason Waddell
##' @export 
#plotDotplot <- function(record) { 
#  
#  if(length(record$lambda) > 0) {
#    
#    par(mar = c(0, 0, 0, 0))
#    
#    xRange <- diff(range(record$lambda)) + 0.5
#    yMax <- max(record$value)
#    
#    xlim <- c(3.25 - 0.24 * xRange,
#        max(record$lambda) + 0.26)
#    ylim <- c(-0.5 - 0.22 * yMax, 
#        max(3, max(record$value)*1.11))
#    blankPlot(xlim = xlim, 
#        ylim = ylim)
#    rect(xlim[1], ylim[1], xlim[2], ylim[2], lwd = 4)
#    
#    text(x = mean(range(record$lambda)), y =  -0.5 - 0.135 * yMax, 
#        labels = "Lambda")
#    text(x = 3.25 - 0.15 * xRange, y =  max(max(record$value)*0.5, 1), 
#        labels = "Piece \nSize")
#    
#    # xAxis
#    segments(x0 = min(record$lambda) - 0.25, y0 = -0.5, 
#        x1 = max(record$lambda) + 0.25,
#        lwd = 2, col = gray(0.2))
#    ticks <- unique(record$lambda)
#    cexX <- 0.9
#    if(max(record$lambda) > 6.5)
#      cexX <- 0.7
#    if(max(record$lambda) > 7)
#      cexX <- 0.6
#    for(i in 1:length(ticks)) {
#      segments(x0 = ticks[i], 
#          y0 = -0.5, y1 = -0.5 - 0.015 * yMax,
#          lwd = 2)
#      text(x = ticks[i], y = -0.5 - 0.025 * yMax,
#          adj = c(0.5, 1), labels = ticks[i], 
#          font = 2, cex = cexX)
#    }
#    
#    # yAxis
#    segments(x0 = 3.25,
#        y0 = -0.5, y1 = max(2.8, max(record$value) * 1.05), col = gray(0.2))
#    ticks <- seq(0, 20, by = 2)[which(seq(0, 20, by = 2) 
#                <= max(2.8, max(record$value)))]
#    for(i in 1:length(ticks)) {
#      segments(x0 = 3.25 - 0.02 * xRange, 
#          y0 = ticks[i], x1 = 3.25,
#          lwd = 2)
#      text(x = 3.25 - 0.04 * xRange,
#          y = ticks[i], adj = c(1, 0.5), labels = ticks[i], 
#          font = 2)
#    }
#    
#    segments(x0 = min(record$lambda), x1 = max(record$lambda), 
#        y0 = min(record$lambda), y1 = max(record$lambda), 
#        col = grey(0.8), lwd = 2)
#    
#    x <- record$lambda
#    y <- record$value
#    
#    xVals <- sort(unique(x))
#    totalSpread <- 0.08
#    
#    globalX <- numeric(length = length(y))
#    meanVec <- rep(0, length(xVals))
#    
#    for(i in 1:length(xVals)){
#      
#      index <- which(x == xVals[i])
#      vec <- y[index]; 
#      meanVec[i] <- mean(vec)
#      
#      maxRep <- max(table(vec))	
#      xLeft = if(maxRep == 2) xVals[i] - totalSpread/4 else xVals[i] - totalSpread/2
#      xRight = if(maxRep == 2) xVals[i] + totalSpread/4 else xVals[i] + totalSpread/2
#      center <- (xLeft+xRight)/2 
#      span <- abs(xRight - xLeft)
#      space <- span/(maxRep-1)	
#      
#      xLocations <- numeric(length = length(vec))
#      
#      for(k in seq_along(table(vec))){
#        numRep <- table(vec)[k]
#        tempLoc <- findLocations(n = numRep, space = space, center = center)
#        
#        tempIndex <- which(vec == sort(unique(vec))[k])
#        
#        for(j in seq_along(tempIndex))
#          xLocations[tempIndex[j]] <- sort(tempLoc)[j]	
#        
#      }
#      
#      globalX[index] <- xLocations
#    }
#    
#    points(x = xVals, y = meanVec, type = "l", lwd = 2)
#    points(x = globalX, y = y, pch = 19, col = record$color, cex = 1.5)
#  } else {
#    par(mar = c(0, 0, 0, 0))
#    blankPlot(xlim = c(0, 1), 
#        ylim = c(0, 1))
#    rect(0, 0, 1, 1, lwd = 4)
#  }
#}



##' (defunct) plot the record of sampled ratio values (no longer works)
##' @param record 
##' @return 
##' @author Jason
##' @export
#plotRecord <- function(record) {
#  
#  par(mar = c(0, 0, 0, 0))
#  if(length(record$lambda) > 0) {
#    maxY <- max(3, max(record$ratio, na.rm=TRUE))
#    blankPlot(xlim = c(-0.07, 1), ylim = c(0, maxY + 0.5))
#    segments(x0 = 0, y0 = 1, x1 = 1, lwd = 2, col = "grey")
#    segments(x0 = 0, y0 = 0, y1 = 3, lwd = 3)
#    ticks <- c(0, 0.5, 1, 2, 3)
#    
#    text(x = 0.5, y = maxY + 0.4, labels = "x_i / E(x)", font = 2, 
#        cex = 2)
#    
#    for(i in 1:length(ticks)) {
#      segments(x0 = -0.02, y0 = ticks[i], x1 = 0, lwd = 2)
#      text(x = -0.04, y = ticks[i], adj = c(1, 0.5), labels = ticks[i], 
#          font = 2)
#    }
#    
#    record$xVal <- record$id / (max(record$id) + 1)
#    points(x = record$xVal, y = record$runningMean, pch = 19,
#        lwd = 2, type = "l")
#    points(x = record$xVal, y = record$ratio, pch = 19, 
#        col = as.character(record$color), 
#        cex = 2)
#  }
#  
#}



inst/ui/ui.R
fluidPage(
    
    includeCSS(file.path("css", "styles.css")),
    
    tags$head(tags$title("Poissontris")), # window header
    
    h1("poissontris - a shiny app", span("(poisson-sampled block sizes)"),
        img(src = "logo_text.png", style = "float:right; width:240px;")        
    ),
    
    
    # keyboard inputs
    tags$script('
            var leftCounter = 0;
            var downCounter = 0;
            var rightCounter = 0;
            var clockCounter = 0;
            var counterCounter = 0;
            $(document).on("keydown", function (e) {
            switch (e.which) {
            case 40:
            Shiny.onInputChange("down", 50);
            Shiny.onInputChange("downCounter", ++downCounter);
            break;
            case 39:
            Shiny.onInputChange("right", 50);
            Shiny.onInputChange("rightCounter", ++rightCounter);
            break;
            case 37:
            Shiny.onInputChange("left", 50);
            Shiny.onInputChange("leftCounter", ++leftCounter);
            break;
            case 83:
            Shiny.onInputChange("counter", ++counterCounter);
            break;
            case 68:
            Shiny.onInputChange("clock", ++clockCounter);
            break;
            case 38:
            Shiny.onInputChange("clock", ++clockCounter);
            break;
            case 70:
            Shiny.onInputChange("swap", Math.random(1));
            break;
            case 13:
            Shiny.onInputChange("startGame", Math.random(1));
            break;
            }});
            '), 
    tags$script('
            $(document).on("keyup", function (e) {
            switch (e.which) {
            case 40:
            Shiny.onInputChange("down", 30);
            break;
            case 39:
            Shiny.onInputChange("right", 30);
            break;
            case 37:
            Shiny.onInputChange("left", 30);
            break;
            }});
            '), 
    
    # TODO column around this
    fluidRow(
        column(3, offset = 2,
            fluidRow(
                column(12, 
                    uiOutput("s1", style="max-height:80%;"), style="position:absolute;"),
                column(12, 
                    uiOutput("s2", style="max-height:80%;"), style="position:absolute;"),
                column(12, 
                    uiOutput("s3", style="max-height:80%;"), style="position:absolute;"), 
                column(12, 
                    uiOutput("s4", style="max-height:80%;"), style="position:absolute;")     
            )), 
        
        column(2,  
            
            h3(HTML("&lambda;:"), textOutput(outputId = "lambda", inline = TRUE)),
            h3(textOutput(outputId = "leftCounter")),
            h3("Score: ", textOutput(outputId = "score", inline = TRUE)),
            h3(icon("trophy", class = "fa-2x"), textOutput("highScore", inline = TRUE)),
            h4("Next Piece: ", style = "margin-top: 50px;"), 
            uiOutput("nextPiece", style="width: 170px; height: 170px;"),
            h4(icon("gamepad", class = "fa-2x"), "Controls: "),
            uiOutput("controls", style="width: 170px; height: 170px;"),   
            
            
            conditionalPanel("input.showButtons % 2 == 1", 
                
                actionButton(inputId = "left", label = ""),
                actionButton(inputId = "leftCounter", label = ""),
                actionButton(inputId = "down", label = ""),
                actionButton(inputId = "downCounter", label = ""),
                actionButton(inputId = "right", label = ""), 
                actionButton(inputId = "rightCounter", label = ""), 
                actionButton(inputId = "counter", label = ""), 
                actionButton(inputId = "clock", label = ""), 
                actionButton(inputId = "clicks", label = ""),
                actionButton(inputId = "swap", label = ""), 
                actionButton(inputId = "startGame", label = "")
            
            )
        ), 
        column(4, 
            htmlOutput(outputId = "pdfTitle", container = h4),
            uiOutput("pdf", style="height: 300px; width: 450px;"), 
            actionButton(inputId = "showPc", label = "Show PC Controls"),
            conditionalPanel("input.showPc % 2 == 1", 
                
                h3(icon("arrow-left", class = "fa-2x"), icon("arrow-down", class = "fa-2x"),
                    icon("arrow-right", class = "fa-2x"),
                    ": move left, down, right"),
                h3(icon("arrow-up", class = "fa-2x"), ": rotate clockwise"),
                h3("s : rotate counterclockwise"),
                h3("f : swap pieces")
            
            )
        
        )
    
    )



)

